python/lsst/daf/butler/_formatter.py

```
@@ -51,12 +63,1034 @@
+    """
+
+    unsupported_parameters: ClassVar[Set[str] | None] = frozenset()
+    """Set of read parameters not understood by this `Formatter`. An empty set
+    means all parameters are supported.  `None` indicates that no parameters
+    are supported. These param (`frozenset`).
+    """
+
+    supported_write_parameters: ClassVar[Set[str] | None] = None
+    """Parameters understood by this formatter that can be used to control
+    how a dataset is serialized. `None` indicates that no parameters are
+    supported."""
+
+    default_extension: str | None = None
+    """Default extension to use when writing a file.
+
+    Can be `None` if the extension is determined dynamically. Use the
+    `get_write_extension` property to get the actual extension to use.
+    """
+
+    supported_extensions: ClassVar[Set[str]] = frozenset()
+    """Set of all extensions supported by this formatter.
+
+    Any extension assigned to the ``default_extension`` property will be
+    automatically included in the list of supported extensions.
+    """
+
+    can_read_from_uri: bool = False
+    """Declare whether `read_from_uri` is available to this formatter."""
+
+    can_read_from_stream: bool = False
+    """Declare whether `read_from_stream` is available to this formatter."""
+
+    can_read_from_local_file: bool = False
+    """Declare whether `read_from_file` is available to this formatter."""
+
+    def __init__(
+        self,
+        file_descriptor: FileDescriptor,
+        *,
+        ref: DatasetRef,
+        write_parameters: Mapping[str, Any] | None = None,
+        write_recipes: Mapping[str, Any] | None = None,
+        # Compatibility parameters. Unused in v2.
+        **kwargs: Any,
+    ):
+        if not isinstance(file_descriptor, FileDescriptor):
+            raise TypeError("File descriptor must be a FileDescriptor")
+
+        self._file_descriptor = file_descriptor
+
+        if ref.isComponent():
+            # Raise or convert to composite ref?
```

I think I lean towards raising, since that's what the docs imply to me and it seems like a bug in higher-level code if a component gets here and it might be confusing to that code if the component is silently ignored.

---

Turns out that for composite disassembly we pass in the component so it can't raise.